metaRNA-seq
===========
Custom scripts for creating and annotating multispecies bacterial metagenomes

Copyright (c) 2013 Peter A. Jorth, Keith H. Turner, Pinar Gumus, Nejat Nizam,
Nurcan Buduneli, Marvin Whiteley

The scripts contained herein can be used to assemble .gff and .fna files
representing a multispecies bacterial "metaorganism" for bioinformatic analyses.
Brief descriptions of the scripts follow, but also see the scripts themselves
and the usage instructions on the command line for information about how to
use them. Please direct any questions on their use or construction to
Keith H. Turner (khturner at utexas.edu).


GenomeMerge.pl
===========
This program takes a tab-separated list of bacterial genome directory names
(find at ftp://ftp.ncbi.nih.gov/genbank/genomes/Bacteria/) and the associated
accession numbers and joins the .fna and the .gff files to make new .fna and
.gff files specifying one "metaorganism" encompassing all of them.

usage: ./GenomeMerge.pl (genomes).txt (out_pfx)

Arguments:
(genomes).txt\tA tab-separated list of directory names and accession numbers
(out_pfx)\tThe prefix you want used for the output files

Dependencies:
-Net::FTP (CPAN)


PullEC.pl
===========
This program takes a GFF file from GenomeMerge.pl on STDIN, searches
http://www.kegg.jp for the locus_tag and returns a new GFF file with EC and
Kegg pathway information on STDOUT.

usage: ./PullEC.pl < (in).gff > (out.gff)

Dependencies:
-wget (Unix)


HOMDpull.sh
===========
This script takes a .gbk file from a dynamically-annotated genome on the Human Oral
Microbiome Database (http://www.homd.org/index.php?&name=seqDownload&type=G)
with prefix (prefix), generates a .fna file of the DNA sequence and a .gff file
with EC number (obtained from the associated _swisstopmatch.xls file) and KEGG
pathway annotations (using the same prefix provided).

usage: ./HOMDpull.sh (prefix)

Requires: (prefix).gbk and (prefix)_swisstopmatch.xls from HOMD

Dependencies:
-bp_seqconvert.pl (BioPerl)
-HOMD_GenomeMerge.pl (this package)


HOMD_GenomeMerge.pl
===========
This is a suite of utilities used in a pipeline to merge genomes from the
Human Oral Microbiome Database (http://www.homd.org).

Usage: ./HOMD_GenomeMerge.pl (command) [arg]

(command) is a required argument that has one of the following values:
GFF_Cleaner - takes a GFF file generated from a GenBank file from the HOMD by 
   bp_genbank2gff on STDIN and parses and relabels the 9th field to match the
   GFF file notation as in GenBank proper on STDOUT.
Get_EC      - takes an (ID, SwissProt Accession No.) pair on STDIN, searches
   UniProt for the accession number and returns (ID, EC number) on STDOUT
   (provide genome prefix at [arg]).
EC_Pathways - takes an (ID, EC number) pair on STDIN, searches http://www.kegg.jp
   for the EC number and returns (ID, EC number, Kegg pathway information) on
   STDOUT (provide genome prefix at [arg]).
Tag_GFF     - reads (ID, EC number, Kegg pathway information) from file [arg] and
   prints (EC number, Kegg pathway information) on the correct line of the GFF
   file read from STDIN on STDOUT.
GFF_Sort     - takes a GFF file generated by our HOMD annotation pipeline on STDIN,
   sorts it by (contig)-(position) and returns it on STDOUT.

Dependencies:
-URI::Escape (CPAN)
-wget (Unix)